examples/fluids/src/matops.c

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2017-2023, Lawrence Livermore National Security, LLC and other CEED contributors.
+// All Rights Reserved. See the top-level LICENSE and NOTICE files for details.
+//
+// SPDX-License-Identifier: BSD-2-Clause
+//
+// This file is part of CEED:  http://github.com/ceed
+
+#include "../include/matops.h"
+
+#include "../navierstokes.h"
+
+// -----------------------------------------------------------------------------
+// Setup apply operator context data
+// -----------------------------------------------------------------------------
+PetscErrorCode SetupMatopApplyCtx(MPI_Comm comm, DM dm, Ceed ceed, CeedOperator op_apply, CeedVector x_ceed, CeedVector y_ceed, Vec X_loc,
+                                  MatopApplyContext op_apply_ctx) {
+  PetscFunctionBeginUser;
+
+  op_apply_ctx->comm   = comm;
+  op_apply_ctx->dm     = dm;
+  op_apply_ctx->x_ceed = x_ceed;
+  op_apply_ctx->y_ceed = y_ceed;
+  op_apply_ctx->op     = op_apply;
+  op_apply_ctx->ceed   = ceed;
+
+  if (X_loc) {
+    op_apply_ctx->X_loc = X_loc;
+    PetscCall(VecDuplicate(X_loc, &op_apply_ctx->Y_loc));
+  } else {
+    op_apply_ctx->X_loc = op_apply_ctx->Y_loc = NULL;
+  }
+
+  PetscFunctionReturn(0);
+}
+
+// -----------------------------------------------------------------------------
+// This function returns the computed diagonal of the operator
+// -----------------------------------------------------------------------------
+PetscErrorCode MatGetDiag(Mat A, Vec D) {
```

Comment:
As written, this looks like a PETSc function. Note that we have `MatGetDiagonal_NS_IJacobian` already.

Reply:
So are you thinking to replace the `Mat*_NS_IJacobian` functions with this generic version of MatShell? Looks like they're effectively equivalent. And I think it makes sense to keep it generic.

Reply:
I see no harm in merging so long as there are no custom semantics to deal with. But don't use a name that looks like a public PETSc function.

Reply:
Gotcha. I just took it from `examples/petsc`.

Reply:
I guess `MatGetDiag_Ceed` would work better.

Reply:
I'll leave replacing the IJacobian functions with the generic equivalents for another PR. I've created an issue here #1142.